### ISSUE TITLE
[BUG]: NeoGeo Battle Coliseum [SLES-54395] - Horrible load times at default EE cycle rate

### ISSUE THREAD

**Issue by Silanda** [PCSX2/pcsx2#11624]
### Describe the Bug

As per the title. The load times in this game are extremely long at default settings, but seem pretty much in line with expectations if EE cycle rate is set to 75%

I don't believe this has always happened, but I haven't played the game in so long that I can't be sure when it became an issue.

### Reproduction Steps

Start the game and enter arcade mode. For reference, I selected Shiki and Mai as my characters. 

The initial load is longer than it should be, but some of the mid fight loads are so long that I wonder if it's stuck in some kind of loop.

### Expected Behavior

Load times aren't the shortest on actual hardware, but are reasonable. Seconds, not minutes. 

### PCSX2 Revision

v2.1.28

### Operating System

Windows 11

### CPU

Ryzen 3700x

### GPU

RTX 4070ti

### GS Settings

Default

### Emulation Settings

Default

EE cycle rate at 75% alleviates the issue.

**Comment by Immersion95** [comment]
The japanese version is the same, loading times are horrendous. But honestly it was like that on my PS2 with original disc. The US version is the only version that was fixed as it was released in 2007 instead of 2006 for the other versions. The load times are much faster there.

**Comment by Immersion95** [comment]
I can reproduce this with the europe version with PCSX2, the loading times are horrendous.

**Comment by StillPlay2** [comment]
> Small correction about the different versions.
> 
> The US version works fine in PCSX2 and on a real PS2/DVD. The slow loading issue seems to affect the Japanese and European versions.
> 
> In Arcade mode, the Japanese and European versions can sometimes take around one minute to load between fights in PCSX2.
> 
> European version (DVD), short loading time: https://www.youtube.com/watch?v=1bntCnE2BIY
> 
> I also tested the Japanese version on a real PS2 through Open PS2 Loader, and the loading is very fast there.
> 
> Video from someone else, Japanese version, short loading time: https://www.youtube.com/watch?v=WmfzyxAdEqM
> 
> I have not found a real japanese PS2/DVD test yet, so I cannot compare directly with the original disc drive. But the Japanese version is fast on real PS2 through OPL, and the EU version is fast in real PS2/DVD, so the one-minute loading time does not look normal.
> 
> The 75% EE cycle rate workaround helps.
> 
> Fast CDVD doesn't help much......but .....I also made a [small experimental PCSX2 fork](https://github.com/Immersion95/pcsx2/actions/runs/25851597232) where Fast CDVD uses x4-x16, OPL uses x4. This does not patch the game code or change the BCCC logic; it only changes the CDVD timing/latency in the emulator.
> 
> With that fork/build, the Japanese and European versions load as fast as, or even faster than the Japanese version on real PS2/OPL.
> 
> So this seems to point more toward a CDVD timing/latency issue in PCSX2 than normal game behavior.
> 
> ## Technical note for debugging
> The issue seems to happen around the BCCC battle loading callback.
> 
> For the Japanese version, the relevant EE area is:
> 
> BCCC callback: 0x0032E110 status check: 0x0032E148 retry branch: 0x0032E150 retry path: 0x0032E1E0 read restart: around 0x0032E214
> 
> For the European version, the equivalent EE area is different because the executable layout is different:
> 
> BCCC status/retry area: 0x00200BA0 - 0x00200BB4 finish path: around 0x00200C80 wait/return path: around 0x00200D98
> 
> When the long loading happens, the game appears to retry the same read repeatedly instead of waiting/checking again after a temporary not-ready status.
> 
> This may help point to where the CDVD timing/status behavior becomes visible in PCSX2.
> 
> ## Temporary patches
> I will also attach my test patches for the Japanese and European versions. [HERE.](https://github.com/user-attachments/files/27749567/Patches.zip)
> 
> They should be seen as a temporary workaround for PCSX2, not as the real emulator fix. They are not needed for OPL, because the game already loads very fast there.
> 
> They may still help a lot on a real PS2 with the original DVD drive + japanese version, since the issue seems related to slower/DVD-like loading timing. I have not tested that case yet, but that is where the patches might make the most sense outside PCSX2.
> 
> For PCSX2, they are a good workaround if you do not want to use the 75% EE cycle rate workaround or a custom PCSX2 build. It is not as fast as OPL, but patch + Fast CDVD is already very close to the US version.
> 
> ## TLDR
> Japanese and European versions can have around one-minute loading between fights in PCSX2.
> 
> The Japanese version is fast on real PS2 through OPL, but I have not found a real PS2/DVD test yet. The European version is fast on real PS2/DVD.
> 
> Fast CDVD doesn't help much, an OPL-like Fast CDVD (x4) makes the loading match or even slightly beat OPL speed.
> 
> The attached patches are only temporary workarounds, but they are useful for PCSX2 users who do not want to use 75% EE cycle rate or a custom build.

I tried your FastCDVD multiplier option fork & it does work wonders in a few games where timing's is of importance.
Maybe there could be a option in PCSX2 Master.

**Comment by Ziemas** [comment]
Keep in mind that OPL re-implements the cdvd api, so it doesn't really tell us anything about the behaviour of the game when actually running from DVD.

**Comment by Immersion95** [comment]
> Keep in mind that OPL re-implements the cdvd api, so it doesn't really tell us anything about the behaviour of the game when actually running from DVD.

Yes that's why I can't confirm anything on the Japanese version on DVD. However there's a video of the European version https://www.youtube.com/watch?v=1bntCnE2BIY which confirms that the bug isn't present on real HW+DVD.

The loading between battles last mere seconds there unlike in PCSX2 which sometimes seem to hang.

**Comment by StillPlay2** [comment]
FastCDVD x4 fixes the osd in audio cd mode going out of sync

**Comment by StillPlay2** [comment]
Interestingly FastCDVD x4 also fixes NHL Hitz 2002 (PAL) intro out of sync.
A PR could be made to see if it fixes other issues related to timings

**Comment by refractionpcsx2** [comment]
Hacking up the CDVD speed isn't really fixing the problem, it needs to be looked in to properly. It might not even be a CDVD problem, it could be indicative that other things are running too fast.  IPU runs instantly (minus the DMA), so that could be a factor.

**Comment by StillPlay2** [comment]
> Hacking up the CDVD speed isn't really fixing the problem, it needs to be looked in to properly. It might not even be a CDVD problem, it could be indicative that other things are running too fast. IPU runs instantly (minus the DMA), so that could be a factor.

Understandable & very true but it could be a starting step to fixing the underlying problem as it seems to fix a myriad of games, but why does it fix it?

**Comment by JordanTheToaster** [comment]
> Understandable & very true but it could be a starting step to fixing the underlying problem as it seems to fix a myriad of games, but why does it fix it?

Investigate the games and find out.

**Comment by Immersion95** [comment]
Can you please remove the "Replicated on Real HW" tag ? Thanks :).

This is a video of the EU version on DVD and it doesn't have the long loading times between battles : https://www.youtube.com/watch?v=1bntCnE2BIY

**Comment by Immersion95** [comment]
## Updated findings (Recap)

The Japanese and European versions can sometimes take around one minute, or appear to load indefinitely, between fights in PCSX2.

This is not normal game behavior:

* European version on a real PS2/DVD, short loading time:
  https://www.youtube.com/watch?v=1bntCnE2BIY
* Japanese version on a real PS2 through OPL, short loading time:
  https://www.youtube.com/watch?v=WmfzyxAdEqM

I have not found a Japanese real-PS2 test using the original DVD drive yet, so that exact case remains unconfirmed.

The 75% EE cycle-rate workaround helps significantly.

Standard Fast CDVD does not help much. I also made an experimental PCSX2 fork where Fast CDVD uses more aggressive CDVD speed and latency settings, while the OPL preset uses x4:

~~https://github.com/Immersion95/pcsx2/actions/runs/25866150307~~

This fork does not patch the game code or change the BCCC callback. It only changes CDVD timing and latency inside PCSX2.

With that build, the Japanese and European versions load as fast as, or sometimes faster than, the Japanese version running through OPL.

This shows that the issue is strongly affected by PCSX2’s CDVD and EE/IOP timing, but the newer executable analysis also found a bug in the game’s retry logic.

## Cause

After an asynchronous disc read, the game calls `sceCdGetError()`.

Its result can mean:

```text
-1  = the status could not be checked yet
 0  = no error
>0  = real CDVD error
```

The Japanese and European versions treat every non-zero result as a failed read:

```c
if (sceCdGetError() != 0)
    restart_the_read();
```

When the status temporarily returns `-1`, the game restarts the same read instead of waiting. Depending on EE, IOP, SIF and CDVD timing, this can repeat many times.

Higher CDVD speed or a 75% EE cycle rate reduces the timing window, which explains why those workarounds help.

## Why the US version behaves differently

The US version still contains similar CDVD retry code, but it also has a separate PreCache loader for 46 common fight files.

Those files are copied from data embedded in the executable instead of being read from the disc. This avoids the problematic CDVD path during many between-fight loads.

## Updated patches

The new Japanese and European patches distinguish a temporary unavailable status from a real disc error:

```c
if (error < 0)
    return;          // wait and check again

if (error > 0)
    retry_read();    // real CDVD error

finish_loading();    // success
```

[Patches](https://github.com/PCSX2/pcsx2_patches/pull/762)

These patches are game-specific workarounds, not a final PCSX2 core fix.

A proper emulator-side fix would require investigating the timing and ordering between EE execution, IOP execution, SIF RPC completion and CDVD command completion.

## TL;DR

* JP and EU can repeatedly restart the same disc read in PCSX2.
* The game mistakes a temporary `sceCdGetError() == -1` result for a real read failure.
* A 75% EE cycle rate helps.
* ~~My experimental PCSX2 fork also fixes the visible loading time by changing CDVD timing and latency only.~~
* The US version avoids many affected reads through its built-in PreCache loader.
* **The updated patches make JP and EU wait on temporary status failures while preserving retries for real CDVD errors. --> This is the best solution at the moment. I will upload it in the official patches database so everyone can enable it without fuss.**
